packages/ozone/src/set/service.ts

```diff
@@ -0,0 +1,203 @@
+        sql<number>`coalesce(sv.count, 0)`.as('setSize'),
+      ])
+      .limit(limit)
+
+    if (namePrefix) {
+      qb = qb.where('s.name', 'like', `${namePrefix}%`)
+    }
+
+    if (cursor) {
+      if (sortBy === 'name') {
+        qb = qb.where('s.name', '>', cursor)
+      } else {
+        qb = qb.where(`s.${sortBy}`, '>', new Date(cursor))
+      }
+    }
+
+    qb = qb.orderBy(`s.${sortBy}`, 'asc')
+
+    const sets = await qb.execute()
+    const lastItem = sets.at(-1)
+
+    return {
+      sets,
+      cursor: lastItem
+        ? sortBy === 'name'
+          ? lastItem?.name
+          : lastItem?.[sortBy].toISOString()
+        : undefined,
+    }
+  }
+
+  async getByName(
+    name: string,
+    txn: Database,
+  ): Promise<Selectable<OzoneSet> | undefined> {
+    const query = txn.db
+      .selectFrom('ozone_set')
+      .selectAll()
+      .where('name', '=', name)
+
+    return await query.executeTakeFirst()
+  }
+
+  async getSetWithValues({
+    name,
+    limit,
+    cursor,
+  }: {
+    name: string
+    limit: number
+    cursor?: string
+  }): Promise<
+    { set: Selectable<OzoneSet>; values: string[]; cursor?: string } | undefined
+  > {
+    const set = await this.getByName(name, this.db)
+    if (!set) return undefined
+
+    let qb = this.db.db
+      .selectFrom('ozone_set_value')
+      .select(['value'])
+      .where('setId', '=', set.id)
+      .limit(limit)
+
+    if (cursor) {
+      qb = qb.where('value', '>', cursor)
+    }
+
+    qb = qb.orderBy('value', 'asc')
+
+    const values = await qb.execute()
+
+    return {
+      set,
+      values: values.map((v) => v.value),
+      cursor: values.at(-1)?.value,
+    }
+  }
+  async upsert(
+    { name, description }: Pick<OzoneSet, 'name' | 'description'>,
+    txn: Database,
+  ): Promise<Selectable<OzoneSet>> {
+    const query = txn.db
+      .insertInto('ozone_set')
+      .values({
+        name,
+        description,
+        updatedAt: new Date(),
+      })
+      .onConflict((oc) =>
+        oc.column('name').doUpdateSet({
+          description,
+          updatedAt: new Date(),
+        }),
+      )
+      .returningAll()
+
+    return await query.executeTakeFirstOrThrow()
+  }
+
+  async addValues(
+    setId: number,
+    values: string[],
+    txn: Database,
+  ): Promise<void> {
+    const query = txn.db
+      .insertInto('ozone_set_value')
+      .values(values.map((value) => ({ setId, value })))
+      .onConflict((oc) => oc.columns(['setId', 'value']).doNothing())
+
+    await query.execute()
+
+    // Update the set's updatedAt timestamp
+    await txn.db
+      .updateTable('ozone_set')
+      .set({ updatedAt: new Date() })
+      .where('id', '=', setId)
+      .execute()
+  }
+
+  async removeValues(
+    setId: number,
+    values: string[],
+    txn: Database,
+  ): Promise<void> {
+    const query = txn.db
+      .deleteFrom('ozone_set_value')
+      .where('setId', '=', setId)
+      .where('value', 'in', values)
```

Comment:
This will fail with a sql syntax error if `values` is empty.

Reply:
we have a check in `/set/remove.ts` to ensure that values is not empty before calling this. added a check in this method too.